app/controllers/pipeline_viz_controller.rb

```diff
@@ -37,4 +63,107 @@ def show
              }, status: status)
     end
   end
+
+  def stage_node_scaffolding(stage_dag_json)
```

Comment:
These and below should private... (but we should move it to the service anyway).